composer/trainer/trainer.py

```diff
@@ -666,16 +666,16 @@ def _train_loop(self) -> None:
                         assert train_metrics is not None
                         self._compute_and_log_metrics(train_metrics, is_train=True, is_batch=True)
 
-                    self.engine.run_event(Event.BATCH_END)
-
-                    for scheduler in state.schedulers:
-                        scheduler.step(interval='batch')  # type: ignore
-
                     state.timer.on_batch_complete(
                         samples=int(num_samples_in_batch.item()),
                         tokens=int(num_tokens_in_batch.item()),
                     )
 
+                    self.engine.run_event(Event.BATCH_END)
+
+                    for scheduler in state.schedulers:
+                        scheduler.step(interval='batch')  # type: ignore
```

Comment:
Looks like batch-level schedulers are incremented after the `BATCH_END` event, but epoch-level schedulers are incremented before `EPOCH_END`? Is that an oversight? Not sure why this change is necessary.

Reply:
Good point; schedulers should be incremented after the timer is incremented. Schedulers should step before the end event so any checkpoint will capture the correct scheduler state.